### DISCUSSION TITLE
In theory, How large files can be used smoothly?

### DISCUSSION THREAD

**Discussion by WingDust** [lapce/lapce#2585] [Q&A]
In theory, How large files can be used smoothly? 1G,2G?

**Comment by dd-dreams** [comment]
What do you mean? Large files are handled as any other file.

**Reply by rbharathkumar** [reply]
My guess is that the original poster wants to understand the ability of lapce to handle large files without hangups or crashes or any other feature getting disabled (for ex certain editors disable find and replace in large files in excess of 5GB).

**Reply by dd-dreams** [reply]
I believe lapce is using the same algorithms for every file to try and make them run smoothly disregarding of the size.
The search and replace for example would probably be the same but just take a longer time. I don't know why certain IDE's disable certain features, it's up to the user to decide if he wants to wait or not.

**Comment by felixsanz** [comment]
lapce hangs on my system opening a 40 MB file...

**Comment by panekj** [comment]
In theory, there is not specific size

**Reply by felixsanz** [reply]
But in practice, there is. Even opening a one-line JSON file with 1MB or so, it hangs

**Comment by MinusGix** [comment]
There's several current problems.  
For single absurdly long lines, there's multiple things.  
The most obvious is 'creating a very big text layout'. Currently our logic would recreate that pretty often, since most lines are very cheap to construct. There's also other line-specific logic that I think assumes operating on lines is cheap, when it isn't necessarily.  
- Simpleish solution: have smarter cache clearing so that they are only recreated when they're actually directly edited
  - This would help when you're not editing that line specifically, but it would still lag
- Proper solution: have some way of applying edits to a text layout, so that it does not need to recreate the entire thing when it is edited. Cosmic-text sortof has this - decent chances that's the intended method based on their examples - but that hasn't been how we're editing it currently.
  - This is more complex with more room for bugs, but would help a lot

(currently, I think Lapce HEAD behaves better in some ways due to wrapping + some other changes done for that. The wrapping breaks it apart more. Not enough still, and there seems to be some buggy behavior with single very long lines and scrolling as well as performance)  

Smarter caching would help with bigger files too.

Big enough files, however, have different issues. A multi-gigabyte file has the issue that we're currently loading it all into memory! Obviously excessive.  
- We should have an ability to load files partially
  - Lot of rooms to accidentally have bugs or weird stutters with this kindof thing
  - And doesn't entirely fit how we've implemented the code, but isn't completely infeasible
- Various pieces of the code should be smarter/faster
- Various pieces of the code should simply disable themselves, or process in a more limited fashion, in large enough files
  - Like syntax highlighting is currently just done to the entire file! Even if you could fit 5gb file in memory, all that extra processing might end up producing a gigabyte or two of syntax information. Smarter code that keeps less around + highlighting a more limited portion + ...

Making all of this fast is certainly a goal, but it is also complicated and doesn't come up all-the-time so the overall feature is kinda on the backburner. It comes up often enough to be an annoyance, however, and we should at least do something marginally better like "Hey, you're trying to load a 1gb file, don't do that, it will make Lapce combust!"

As well, I/we hope to add more benchmarking to various pieces of Lapce which would make finding good changes easier - even if they're not the full-fledged problem.
